data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericExpressionEvaluator.java

```diff
@@ -74,4 +76,57 @@ public Boolean isValidFormatExpression(final String format) {
         }
         return true;
     }
+
+    @Override
+    public List<String> extractDynamicKeysFromFormatExpression(final String format) {
+        final List<String> formatExpressionKeys = new ArrayList<>();
+
+        if (format == null) {
+            return formatExpressionKeys;
+        }
+
+        int fromIndex = 0;
+        int position = 0;
+        while ((position = format.indexOf("${", fromIndex)) != -1) {
+            int endPosition = format.indexOf("}", position + 1);
+            if (endPosition == -1) {
```

Comment:
This might be a user error in dynamic expression. Looks like we are ignoring any errors ?

Reply:
This method is not responsible for catching invalid expressions, it will just parse out the valid ones. If you think it should throw, then it can here but it's job is not to validate, just to parse and return

Reply:
Will add some unit tests for validating the expression on startup of the S3 sink in the next PR

Reply:
We can add a test case to cover validation on startup for s3 sink constructor as discussed offline.